luarules/gadgets/unit_dgun_behaviour.lua

```diff
@@ -57,63 +82,119 @@ local function addVolumetricDamage(projectileID)
+		local x, y, z = spGetProjectilePosition(proID)
+		local h = spGetGroundHeight(x, z)
 
 		if y < h + 1 or y < 0 then -- assume ground or water collision
 			-- normalize horizontal velocity
-			local dx, _, dz, speed = Spring.GetProjectileVelocity(proID)
+			local dx, _, dz, speed = spGetProjectileVelocity(proID)
 			local norm = speed / math.sqrt(dx^2 + dz^2)
 			local ndx = dx * norm
 			local ndz = dz * norm
-			Spring.SetProjectileVelocity(proID, ndx, 0, ndz)
+			spSetProjectileVelocity(proID, ndx, 0, ndz)
 
 			groundedDGuns[proID] = true
 			flyingDGuns[proID] = nil
 		end
 	end
 
 	for proID in pairs(groundedDGuns) do
-		local x, y, z = Spring.GetProjectilePosition(proID)
+		local x, y, z = spGetProjectilePosition(proID)
 		-- place projectile slightly under ground to ensure fiery trail
 		local verticalOffset = 1
-		Spring.SetProjectilePosition(proID, x, math.max(Spring.GetGroundHeight(x, z), 0) - verticalOffset, z)
+		spSetProjectilePosition(proID, x, math.max(spGetGroundHeight(x, z), 0) - verticalOffset, z)
 
 		-- NB: no removal; do this every frame so that it doesn't fly off a cliff or something
 	end
+
+	-- Without defining a time to live (TTL) for the DGun, it will live forever until it reaches maximum range. This means it would deal infinite damage to shields until it depleted them.
+	if next(dgunTimeouts) == nil then
+		frameCounter = 0
+	else
+		frameCounter = frameCounter + 1
+
+		for proID, timeout in pairs(dgunTimeouts) do
+			if frameCounter > timeout then
+				spDeleteProjectile(proID)
+				flyingDGuns[proID] = nil
+				groundedDGuns[proID] = nil
+				dgunTimeouts[proID] = nil
+			end
+		end
+	end
 end
 
 function gadget:UnitPreDamaged(unitID, unitDefID, unitTeam, damage, paralyzer, weaponDefID, projectileID, attackerID, attackerDefID, attackerTeam)
 	if dgunWeapons[weaponDefID] and isCommander[attackerDefID] and (isCommander[unitDefID] or isDecoyCommander[unitDefID]) then
 		if isDecoyCommander[unitDefID] then
 			return dgunWeapons[weaponDefID].damages[0]
 		else
-			Spring.DeleteProjectile(projectileID)
-			local x, y, z = Spring.GetUnitPosition(unitID)
-			Spring.SpawnCEG("dgun-deflect", x, y, z, 0, 0, 0, 0, 0)
+			spDeleteProjectile(projectileID)
+			local x, y, z = spGetUnitPosition(unitID)
+			spSpawnCEG("dgun-deflect", x, y, z, 0, 0, 0, 0, 0)
 			local armorClass = UnitDefs[unitDefID].armorType
 			return dgunWeapons[weaponDefID].damages[armorClass]
 		end
 	end
+
 	return damage
 end
+
+local lastShieldFrameCheck = {}
+
+function gadget:ShieldPreDamaged(proID, proOwnerID, shieldEmitterWeaponNum, shieldCarrierUnitID, bounceProjectile, beamEmitterWeaponNum, beamEmitterUnitID, startX, startY, startZ, hitX, hitY, hitZ)
+	if proID > -1 and dgunTimeouts[proID] then
+		local proDefID = spGetProjectileDefID(proID)
+		local shieldEnabledState, shieldPower = spGetUnitShieldState(shieldCarrierUnitID)
+		local damage = WeaponDefs[proDefID].damages[11] or WeaponDefs[proDefID].damages[2]
```

Comment:
```suggestion
		local damage = WeaponDefs[proDefID].damages[Game.armorTypes.shields]
```

Reply:
> I can't comprehend the fragility. Only case I can think of would be if the projectile died and was removed before this calculation could complete. Hasn't happened yet but sounds possible.
> 
The fragility comes from the code breaking if for any reason the contents of that table are not in the same static order they were when the code was written. Using the named keys, rather than integer keys, ensures the code will not break if the table is in a different order.

It also makes the intent clear, as it is impossible for a dev to understand what the integer keys of a table refer to.
```suggestion
		local damage = WeaponDefs[proDefID].damages[Game.armorTypes.shields] or WeaponDefs[proDefID].damages[Game.armorTypes.default]
```